### PR TITLE
STVM, NUPHAR, remove tvm from submodules list, checks pointers are not null.

### PR DESCRIPTION
**Pull request by xadupre**
**Description**:
* Checks pointers are not null to avoid crashes.
* Removes tvm from the list of submodules and use FetchContent to retrieve the sources. It should speed up the build for every other scenarios.

**Motivation and Context**
Help identy why it fails. Speed up CI.

### CODE REVIEW

File: cmake/external/tvm.cmake
Comment (snnn) [inline cmake/external/tvm.cmake:19]: Let's remove NUPHAR.

File: cmake/external/tvm.cmake
Comment (snnn) [inline cmake/external/tvm.cmake:1]: You still need to add nuphar and all its dependencies that will be used in your build to cgmanifest.json.

File: docs/STVM_EP.md
Comment (snnn) [inline docs/STVM_EP.md:38]: Why users need to manually build TVM first? If that, it seems we don't need to clone TVM. We can treat TVM as CUDA SDK. We ask the user provide the install path of the software. 

In whatever case, we still need to track the software with cgmanifest.json.

Comment (xadupre) [reply docs/STVM_EP.md:38]: It comes from PR #10019. The first draft requires to manually build TVM. There is also an ongoing work with PR #10260 (renaming).

**Comment by vvchernov** [comment]
Hello @xadupre, I think it is good idea leave one tvm submodule instead of two but you should know the following things. 1. Nuphar and STVM are the different EPs although they both are based on Apache TVM. They are developed by different teams and use different TVM hash. For STVM I do not think that v0.8.0 is a good choice, we use very fresh commit. Separately it should be agreed with Nuphar team.
2. If you remove tvm submodules it is needed to have TVM elsewhere and build it before ORT. Particularly so-file from TVM is used for whl-package build (onnxruntime-stvm). Moreover build system of ORT is big enough it looks like you need check(update) more places for correct work without tvm submodules.
3. TVM build outside of ORT becames the responsibility of client. But it should be noted that TVM is target-dependent project. With which keys do you plan to build it and use for ORT? There is no universal answer. Our idea was to add cmake keys to build TVM together with ORT and based on keys from ORT, for GPU we use CUDA or Metal or Vulkan, for CPU - LLVM and so on.
4. Please also take into account [PR#10260](https://github.com/microsoft/onnxruntime/pull/10260).

**Comment by xadupre** [comment]
Both NUPHAR and STVM are using TVM but not the same versions (nuphar is using a fork not updated in two years). I used whatever was already checked in. Between 0.5 and 0.8, TVM updated its C++ API, updating TVM for nuphar would require many changes (see #10183). This PR does not remove TVM, it fetches TVM sources only if it is needed. The build does not change for nuphar or any other version. The goal of this PR is let the build unchanged but clone tvm only when it is needed.

**Comment by tmoreau89** [comment]
Thank you @xadupre - I think we can go ahead and merge this PR if it's ready, after which we can go ahead and land the following PR: https://github.com/microsoft/onnxruntime/pull/10260

**Comment by jroesch** [comment]
Makes sense to me as well, thanks for the improvement @xadupre

**Comment by vvchernov** [comment]
Hello @xadupre I recommended you to check build of TVM inside ORT without USE_MICRO key. The latter activates tvmc which allows to work with TVM by commands from shell I do not think that it needs for ORT. We also will check it tommorow on our side

**Comment by xadupre** [comment]
I tried but it was crashing when running "import onnxruntime.providers.stvm". So I wanted to try with the same setting given in the documentation.

**Comment by KJlaccHoeUM9l** [comment]
Hello @xadupre!
This is because after building the project, you have two `libtvm.so` files located in two different directories:
1. `onnxruntime/build/Linux/RelWithDebInfo/_deps/tvm-build/libtvm.so`
2. `onnxruntime/build/Linux/RelWithDebInfo/_deps/tvm-src/python/tvm/libtvm.so`

Both Python packages (`onnxruntime-stvm`, `tvm`) depend on the same `libtvm.so`, so when you `import onnxruntime.providers.stvm`, `libtvm.so` loads from the directory that was involved in the CMake build (the first directory). And after that, with the internal `import tvm`, a second load of `libtvm.so` occurs from the second directory. At this point, an error occurs, because when reloading, the methods from `libtvm.so` are re-registered.

The solution to this problem can be the following ways:
1. Set the `LD_LIBRARY_PATH` variable to one of the specified directories. Then `libtvm.so` will be loaded from one directory.
2. Insert at the beginning of the test script `import tvm`. In this case, TVM will take over the logic of loading `libtvm.so` from the right place and no error will occur.

The second solution has already been implemented in [PR](https://github.com/microsoft/onnxruntime/pull/10241).

Additionally, the work was tested through the built wheel packages. If you do `pip install onnxruntime/build/Linux/RelWithDebInfo/_deps/tvm-src/python/dist/tvm-0.8.0-cp38-cp38-linux_x86_64.whl`, then when you try to `import tvm` there will be an error related to the fact that not `libtvm.so` can be found. This error can also be solved with `LD_LIBRARY_PATH`, but it looks weird. I used to build the wheel package:
```Shell
cd tvm/python
python setup.py bdist_wheel
```
In this case, no problems arose, it is worth investigating in more detail the way the wheel package is built in this PR.

Additionally, TVM seems to be built without the `USE_LLVM` flag. If you look at the `onnxruntime/tools/ci_build/build.py` file (which I use to build the project), you can see that the `-Donnxruntime_USE_LLVM` flag is only enabled when building under `nuphar`, although this is also necessary for `stvm`.

**Review by snnn — DISMISSED** [review]

**Comment by vvchernov** [comment]
Hello @xadupre, we are concerned about possible situation when this PR will be merged due to CI passed, but in fact it will not work for ORT build with TVM or separated TVM build or TVM whl-package build and so on. May be we can help you or can communicate/test to be sure that all things work well before it will be merged?

**Comment by xadupre** [comment]
Hello @vvchernov, the only way to make sure of it is to add a CI job with a unit test. It is working for me with the following steps:

```
python3 ./tools/ci_build/build.py --config Release --skip_tests --build_wheel --parallel  --build_dir ./build/cpu_stvm --use_stvm
export PYTHONPATH=~/github/onnxruntime/build/cpu_stvm/Release/:~/github/onnxruntime/build/cpu_stvm/Release/_deps/tvm-src/python
```

Then a simple script:

```
import tvm  # needs to be imported first
import onnxruntime
# ... the rest of the script
```

Now I need to check if it works too after installing both *.whl built during this build.

**Comment by KJlaccHoeUM9l** [comment]
Hello @xadupre, `import tvm` will be automatically done after [PR#10241](https://github.com/microsoft/onnxruntime/pull/10241) merge.
When building the `.whl` package for `onnxruntime-stvm`, this import will be automatically inserted into `_ld_preload.py`. Thus, `import tvm` will be called at the beginning of the entire chain of imports in onnxruntime.
Therefore, `import tvm` can be removed from your test script.

**Comment by xadupre** [comment]
Hi @KJlaccHoeUM9l, this PR was meged and I built with this one included. I did not work for me. I'll try again.

**Comment by KJlaccHoeUM9l** [comment]
Sorry, it was my mistake in [PR#10241](https://github.com/microsoft/onnxruntime/pull/10241).
Due to active work with a branch in which STVM has already been renamed to TVM, I made a mistake on [this line](https://github.com/microsoft/onnxruntime/blob/4af116649c8f5f6e725ce8b314b7f8e38007f236/setup.py#L238).
For correct work, you need to change this line as follows:
```Python
if package_name == 'onnxruntime-stvm':
```
Maybe it will be faster to make this change within this PR? Or should I open a new PR?

**Comment by xadupre** [comment]
@KJlaccHoeUM9l, I fixed it in this PR. I added a small example to check the provider is working. Is it possible to review it? It fails foe me: https://github.com/microsoft/onnxruntime/pull/10211/files#diff-c6a93b897de7a145372d94463bb7aa6030f3bd3578c50548a261025b6825ade6R53. The inference works but CPUExecutionProvider and StvmExecutionProvider do not give the same output.

**Comment by vvchernov** [comment]
> @KJlaccHoeUM9l, I fixed it in this PR. I added a small example to check the provider is working. Is it possible to review it? It fails foe me: https://github.com/microsoft/onnxruntime/pull/10211/files#diff-c6a93b897de7a145372d94463bb7aa6030f3bd3578c50548a261025b6825ade6R53. The inference works but CPUExecutionProvider and StvmExecutionProvider do not give the same output.

@xadupre which model did you use for test?

**Comment by xadupre** [comment]
> > @KJlaccHoeUM9l, I fixed it in this PR. I added a small example to check the provider is working. Is it possible to review it? It fails foe me: https://github.com/microsoft/onnxruntime/pull/10211/files#diff-c6a93b897de7a145372d94463bb7aa6030f3bd3578c50548a261025b6825ade6R53. The inference works but CPUExecutionProvider and StvmExecutionProvider do not give the same output.
> 
> @xadupre which model did you use for test?

The test has to be fast so I did not use any predefined model. The model is created in the script using onnx functions. It is a linear regression (XA + B).

**Comment by KJlaccHoeUM9l** [comment]
@xadupre, this error is due to the fact that your model does not have fixed tensor shapes. At the moment, STVM EP in automatic mode can only work with models that have fixed shapes.
In order for your test to work correctly, you need to change it as follows:
```Python
X = make_tensor_value_info('X', TensorProto.FLOAT, [1, 2])
A = make_tensor_value_info('A', TensorProto.FLOAT, [2, 2])
B = make_tensor_value_info('B', TensorProto.FLOAT, [1, 2])
Y = make_tensor_value_info('Y', TensorProto.FLOAT, [1, 2])
```

**Comment by xadupre** [comment]
I suggest to raise an exception if the model has dynamic shape and input_shapes is not specified.

File: cmake/CMakeLists.txt
Comment (snnn) [inline cmake/CMakeLists.txt:1363]: So, please also add dmlc-core and dlpack to our cgmanifest.json. In the json file you may add a comment saying they are for TVM.

File: cmake/external/tvm.cmake
Comment (xadupre) [reply cmake/external/tvm.cmake:1]: TVM dependencies are listed here: https://github.com/apache/tvm/blob/main/LICENSE#L204. Some of them are checked in the repository and the sources are difficult to trace (cma). Another comes from another dependency already listed in this file (compiler-rt / llvm, blockingconcurrentqueue.h).

**Review by snnn — APPROVED** [review]
LGTM.  You may merge it and continue to improve the code.

**Comment by prateek9623** [comment]
@xadupre hey can you help me with NUPHAR-TVM cmake install at this PR:https://github.com/microsoft/onnxruntime/pull/8919

**Comment by tmoreau89** [comment]
Great, thank you @xadupre for getting this PR through and @snnn @KJlaccHoeUM9l and @vvchernov for the review.

File: cmake/CMakeLists.txt
Comment (prateek9623) [inline cmake/CMakeLists.txt:1354]: @xadupre  Do we require EXCLUDE_FROM_ALL here, as this prevents "tvm" to be installed when "onnxruntime" is installed?

**Pull request merged**
